ot/gromov/_unbalanced.py

```diff
@@ -0,0 +1,1210 @@
+        two unbalanced optimal transport problems in each BCD iteration.
+    tol_ot : float, optional (default = 1e-7)
+        Tolerance of unbalanced solver for each of the
+        two unbalanced optimal transport problems in each BCD iteration.
+    method_sinkhorn : string, optional (default = "sinkhorn")
+        Method used in POT's `ot.sinkhorn` solver when divergence = "kl" and
+        unbalanced_solver = "scaling". Only support method_sinkhorn = "sinkhorn"
+        and method_sinkhorn = "sinkhorn_log".
+    log : bool, optional (default = False)
+        If True then the cost and 4 dual vectors, including
+        2 from sample and 2 from feature couplings, are recorded.
+    verbose : bool, optional (default = False)
+        If True then print the COOT cost at every multiplier of `eval_bcd`-th iteration.
+
+    Returns
+    -------
+    pi_samp : (n_sample_x, n_sample_y) array-like, float
+        Sample coupling matrix.
+    pi_feat : (n_feature_x, n_feature_y) array-like, float
+        Feature coupling matrix.
+    log : dictionary, optional
+        Returned if `log` is True. The keys are:
+            error : list of L1 norms between the current and previous sample coupling.
+            duals_sample : (n_sample_x, n_sample_y) tuple, float
+                Pair of dual vectors when solving OT problem w.r.t the sample coupling.
+            duals_feature : (n_feature_x, n_feature_y) tuple, float
+                Pair of dual vectors when solving OT problem w.r.t the feature coupling.
+            linear : float
+                Linear part of the cost.
+            ucoot : float
+                Total cost.
+    """
+
+    #############################
+    # Calculate D(pi, a \otimes b) for D = KL and squared L2.
+
+    # Calculate KL(pi, a \otimes b) using
+    # the marginal distributions pi1, pi2 of pi
+    def approx_shortcut_kl(pi, pi1, pi2, a, b):
+        """
+        Implement:
+        < pi, log pi / (a \otimes b) >
+        = <pi, log pi> - <pi1, log a> - <pi2, log b>.
+        """
+
+        res = nx.sum(pi * nx.log(pi + 1.0 * (pi == 0))) \
+            - nx.sum(pi1 * nx.log(a)) - nx.sum(pi2 * nx.log(b))
+        return res
+
+    def div(pi, pi1, pi2, a, b):
+        """
+        Calculate D(pi, a \otimes b).
+        """
+
+        if divergence == "kl":
+            res = approx_shortcut_kl(pi, pi1, pi2, a, b) \
+                - nx.sum(pi1) + nx.sum(a) * nx.sum(b)
+        elif divergence == "l2":
+            res = (nx.sum(pi**2) + nx.sum(a**2) * nx.sum(b**2)
+                   - 2 * nx.dot(a, pi @ b)) / 2
+        return res
+
+    #############################
+    # Support functions for KL and squared L2 between product measures:
+    # Calculate D(mu \otimes nu, alpha \otimes beta).
+    def approx_kl(p, q):
+        return nx.sum(p * nx.log(p + 1.0 * (p == 0))) - nx.sum(p * nx.log(q))
+
+    def kl(p, q):
+        return approx_kl(p, q) - nx.sum(p) + nx.sum(q)
+
+    def product_kl(mu, nu, alpha, beta):
```

Comment:
unfortunately, no because need to redefine nx. However, I think it's better to define out of the function for more clarity, at the cost of defining nx (hopefully very cheap).

Reply:
you could set `nx=None` (backend optimal parameter) as parameter of an independent function `product_kl`, test if it is None you deduce it from mu and nu, otherwise use the one provided as input.

Reply:
agree, already on the way to implement. I've just consulted the support functions in _utils.py